File: whalebot/alert_system.py
# alert_system.py

import logging
import subprocess

from colorama import Fore, Style, init

# Initialize colorama
init(autoreset=True)

NEON_RED = Fore.LIGHTRED_EX
NEON_YELLOW = Fore.YELLOW

class AlertSystem:
    """Handles sending alerts for critical bot events using Termux toast notifications.
    """

    def __init__(self, logger: logging.Logger):
        """Initializes the AlertSystem.

        Args:
            logger: The logger instance to use for logging.
        """
        self.logger = logger

    def send_alert(self, message: str, level: str = "INFO"):
        """Sends an alert using termux-toast.

        Args:
            message: The alert message to display.
            level: The severity level (e.g., "INFO", "WARNING", "ERROR").
                   This can be used to colorize the toast in future versions.
        """
        self.logger.info(f"Attempting to send {level} alert: {message}")
        try:
            # Using subprocess.run to send a toast notification
            # The '-s' flag allows the toast to be shown for a shorter duration
            subprocess.run(
                ['termux-toast', message],
                check=True,
                capture_output=True,
                text=True
            )
            self.logger.info("Termux toast alert sent successfully.")
        except FileNotFoundError:
            self.logger.error(
                f"{NEON_RED}The 'termux-toast' command was not found. "
                f"Please ensure the Termux:API app is installed and that you have run 'pkg install termux-api'.{Style.RESET_ALL}"
            )
        except subprocess.CalledProcessError as e:
            self.logger.error(
                f"{NEON_RED}Failed to send Termux toast notification.{Style.RESET_ALL}\n"
                f"{NEON_YELLOW}Stderr: {e.stderr}{Style.RESET_ALL}"
            )
        except Exception as e:
            self.logger.error(f"{NEON_RED}An unexpected error occurred while sending a toast: {e}{Style.RESET_ALL}")

File: pybit/javabot/utils/alert_system.js
import { execSync } from 'child_process';
import { logger } from '../logger.js';
import chalk from 'chalk';

/**
 * @class AlertSystem
 * @description Provides a system for sending alerts, including logging and Termux toast notifications.
 */
class AlertSystem {
    /**
     * @constructor
     * @description Initializes the AlertSystem and checks for Termux API availability.
     */
    constructor() {
        /**
         * @property {boolean} termux_api_available - Indicates if Termux API is available for toast notifications.
         */
        this.termux_api_available = this._check_termux_api();
    }

    /**
     * @private
     * @method _check_termux_api
     * @description Checks if the `termux-toast` command is available in the system.
     * @returns {boolean} True if `termux-toast` is found, false otherwise.
     */
    _check_termux_api() {
        try {
            execSync('which termux-toast', { stdio: 'pipe' });
            return true;
        } catch (e) {
            logger.warn(chalk.yellow("Termux toast notifications disabled (command not found)."));
            return false;
        }
    }

    /**
     * @method send_alert
     * @description Sends an alert message, logging it and optionally displaying it as a Termux toast.
     * @param {string} message - The alert message to send.
     * @param {string} [level="INFO"] - The severity level of the alert ("INFO", "WARNING", "ERROR").
     * @returns {void}
     */
    send_alert(message, level = "INFO") {
        const colorMap = {
            INFO: chalk.blue,
            WARNING: chalk.yellow,
            ERROR: chalk.red
        };
        const prefixMap = {
            INFO: "ℹ️ ",
            WARNING: "⚠️ ",
            ERROR: "⛔ "
        };

        const color = colorMap[level] || chalk.white;
        const prefix = prefixMap[level] || "";
        logger.log(level.toLowerCase(), `${prefix}${message}`);

        if (this.termux_api_available) {
            try {
                execSync(`termux-toast "${prefix}${message}"`, { timeout: 5000 });
            } catch (e) {
                logger.error(chalk.red(`Termux toast failed: ${e.message}`));
            }
        }
    }
}

export default AlertSystem;